### DISCUSSION TITLE
How to set Job timeout?

### DISCUSSION THREAD

**Discussion by pyk** [dagster-io/dagster#12463] [Q&A] [answered]
We've noticed that some of the job runs are stale after every code servers re-deployment and not continuing the execution:

<img width="1349" alt="Screenshot 2023-02-22 at 14 13 52" src="https://user-images.githubusercontent.com/2213646/220549187-b42ca90d-1e68-4d50-8b4a-f79c5c479973.png">

stale steps:
<img width="1351" alt="Screenshot 2023-02-22 at 14 22 12" src="https://user-images.githubusercontent.com/2213646/220550735-d38b755d-647c-406b-8d75-646c78dab134.png">

currently we have to manually re-execute then delete all stale job runs.

The problem is that this stale job runs will pile up and blocking all the jobs run.

Is there any way to set timeout per job runs? 

If there is no way to set the job timeout, I want to build sensor that do the following:

1. automatically get all jobs run
2. get execution time
3. if execution_time > TIMEOUT: re submit job 

Is it possible to access this dagster info inside the sensor? 

thank you! 🙏

**Comment by clairelin135** [comment]
Hi @pyk. We don't have a built-in way to do this at the moment, but you could do it through a sensor. Something like this:

```
@sensor
def retry_job_if_timeout(context):
    run_coordinator = context.instance.run_coordinator
    in_progress_runs = context.instance.get_runs(
        filters=PipelineRunsFilter(statuses=[PipelineRunStatus.STARTED])
    )
    for run in in_progress_runs:
        stats: PipelineRunStatsSnapshot = context.instance.get_run_stats(run.run_id)
        elapsed_time = stats.end_time - stats.start_time
        if elapsed_time > max_elapsed_seconds:
            # first try to cancel via the run coordinator, which is responsible for cleaning up
            # any compute resources that might be devoted to the run
            is_canceled = (
                run_coordinator
                and run_coordinator.can_cancel_run(run.run_id)
                and run_coordinator.cancel_run(run.run_id)
            )

            if not is_canceled:
                reloaded_run = context.instance.get_run_by_id(run.run_id)
                if not reloaded_run.is_finished:
                    # next mark the run as canceled, even though there might still be compute
                    # that has not been cleaned up.
                    context.instance.report_run_canceled(
                        run, message=f"canceled by cancel_pipeline after {elapsed_time} seconds"
                    )
```

**Reply by pyk** [reply]
@clairelin135 tysm! 🙇

**Reply by tandav** [reply]
This code is using [removed APIs `PipelineRunStatus`](https://github.com/dagster-io/dagster/blob/7730cd0b53efba51ab2e98c8c00e9dbb93c70ff6/MIGRATION.md#removed-api-list)

Also what is `PipelineRunsFilter`?  I cant find it in the repositpory:

```py
$ grep -r --include '*.py' PipelineRunsFilter .
./python_modules/dagster-graphql/dagster_graphql/schema/errors.py:class GrapheneInvalidPipelineRunsFilterError(graphene.ObjectType):
./python_modules/dagster-graphql/dagster_graphql/schema/errors.py:        name = "InvalidPipelineRunsFilterError"
./python_modules/dagster-graphql/dagster_graphql/schema/errors.py:    GrapheneInvalidPipelineRunsFilterError,
./python_modules/dagster-graphql/dagster_graphql/schema/runs.py:    GrapheneInvalidPipelineRunsFilterError,
./python_modules/dagster-graphql/dagster_graphql/schema/runs.py:        types = (GrapheneRuns, GrapheneInvalidPipelineRunsFilterError, GraphenePythonError)
./python_modules/dagster/dagster/_core/storage/dagster_run.py:    old_storage_names={"PipelineRunsFilter"},
```

**Reply by C0DK** [reply]
I'd love to hear what the current version of this is. I'd like to create a timeout of all runs so no run is longer than 10 hours.

**Comment by catalyst17** [comment]
Hello! We are facing a similar issue of stuck jobs that block the queue and result in accumulation of pending `STARTING` jobs

Do I understand correctly that after the discussion above happened, [these features](https://docs.dagster.io/deployment/run-monitoring) were added and the example sensor provided by @clairelin135 should be no longer needed?

**Comment by gibsondan** [comment, marked as answer]
We now have a way to do this natively within Dagster: https://docs.dagster.io/deployment/run-monitoring#general-run-timeouts

**Reply by toddy86** [reply]
@gibsondan is this available at the individual asset/op level, or only at the job level as the docs suggests?

If not available at an individual asset/op level yet, this would be extremely useful. Sometimes, there are jobs that combined should not exceed x-time, but more often we are concerned about the time individual assets/ops take (e.g. asset 1 should not exceed 1 hour, asset 2 in the same job should not exceed 3 hours).

As a secondary question, how does this apply to assets/ops that use auto-materialize? 

<img width="790" alt="image" src="https://github.com/dagster-io/dagster/assets/10559757/3a6c1123-420a-4bb5-9cdd-e6ce7f8c5a9a">

**Comment by gibsondan** [comment]
This is currently only available for jobs. There’s a separate feature
request issue in the dagster repo for op level timeouts

**Reply by TabLand** [reply]
> is this available at the individual asset/op level, or only at the job level as the docs suggests?

Timeouts have been requested under:
- #3666
- #17498
